### PR TITLE
[Train] Mention threading in `TensorflowTrainer` doc

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Even though we allow users to set the number of CPUs per worker, we do not set Tensorflow threading configurations automatically. This PR mentions this explicitly in the docstring.

We may consider setting this automatically in the future, but the main concern is that there are multiple configuration options interacting with another, so just setting them all to the number of CPUs assigned to a worker may not give the best performance. Therefore, we leave it up to the user for now.

For more information, see
https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads
https://www.intel.com/content/www/us/en/developer/articles/technical/maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference.html

Quote from the second link:

> The ConfigProto is used for configuration when creating a session. These two variables control number of cores to use.
> 
> intra_op_parallelism_threads
> This runtime setting controls parallelism inside an operation. For instance, if matrix multiplication or reduction is intended to be executed in several threads, this variable should be set. TensorFlow will schedule tasks in a thread pool that contains intra_op_parallelism_threads threads. As illustrated later in Figure 2, OpenMP* threads are bound to thread context as close as possible on different cores. Setting this environment variable to the number of available physical cores is recommended.
> 
> inter_op_parallelism_threads
> NOTE: This setting is highly dependent on hardware and topologies, so it’s best to empirically confirm the best setting on your workload.
> 
> This runtime setting controls parallelism among independent operations. Since these operations are not relevant to each other, TensorFlow will try to run them concurrently in the thread pool that contains inter_op_parallelism_threads threads. This variable should be set to the number of parallel paths where you want the code to run. For Intel® Optimization for TensorFlow, we recommend starting with the setting '2’, and adjusting after empirical testing.

## Related issue number

Closes https://github.com/anyscale/product/issues/14817

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/tensorflow/tensorflow_trainer.py [hunk 1]
```diff
@@ -45,6 +45,16 @@ def train_loop_per_worker(config: Dict):
...
     :ref:`Ray AIR session methods <air-session-ref>`.
 
+    .. warning::
+        Ray will not automatically set any environment variables or configuration
+        related to local parallelism / threading
+        :ref:`aside from "OMP_NUM_THREADS" <omp-num-thread-note>`.
+        If you desire greater control over TensorFlow threading, use
+        the ``tf.config.threading`` module (eg.
+        ``tf.config.threading.set_inter_op_parallelism_threads(num_cpus)``)
+        at the beginning of your ``train_loop_per_worker`` function.
```
Comment (matthewdeng) [inline python/ray/train/tensorflow/tensorflow_trainer.py:55]: Can we move this to a TensorFlow user guide instead?

Comment (Yard1) [reply python/ray/train/tensorflow/tensorflow_trainer.py:55]: Will that have enough visibility?

Comment (Yard1) [reply python/ray/train/tensorflow/tensorflow_trainer.py:55]: I've put it in both places

**Comment by amogkam** [comment]
Should we add these as args to `TensorflowConfig`? Make it easy for users to specify these and also make it a more explicit part of the API. Then we can log a warning if num_cpus_per_worker>1 in TensorflowTrainer, recommending users to set these.

**Comment by Yard1** [comment]
@amogkam While I like the idea, one thing I need to call out is that we'd burden ourselves with making sure we keep our `TensorflowConfig` up to date with whatever changes Tensorflow makes. Aside from the warning (which is, admittedly, very useful), we would be kind of duplicating what can already be easily done with Tensorflow.

That being said, I think changes to those APIs are unlikely on the TF side, and it's just 2 configuration options, so that should be fine!

Not sure if we want to merge this before the branch cut and then do a followup with `TensorflowConfig`, or do `TensorflowConfig` right away but miss the cut

**Review by amogkam — APPROVED** [review]

**Pull request merged**